src/chainstate/stacks/boot/mod.rs

```diff
@@ -249,39 +603,85 @@ impl StacksChainState {
     /// If an address appears in `addresses` multiple times, then the address's associated amounts
     ///   are summed.
-    pub fn make_reward_set(
-        threshold: u128,
-        mut addresses: Vec<(PoxAddress, u128)>,
-    ) -> Vec<PoxAddress> {
+    pub fn make_reward_set(threshold: u128, mut addresses: Vec<RawRewardSetEntry>) -> RewardSet {
         let mut reward_set = vec![];
+        let mut missed_slots = vec![];
         // the way that we sum addresses relies on sorting.
-        addresses.sort_by_key(|k| k.0.bytes());
-        while let Some((address, mut stacked_amt)) = addresses.pop() {
+        addresses.sort_by_key(|k| k.reward_address.bytes());
+        while let Some(RawRewardSetEntry {
+            reward_address: address,
+            amount_stacked: mut stacked_amt,
+            stacker,
+        }) = addresses.pop()
+        {
+            let mut contributed_stackers = vec![];
+            if let Some(stacker) = stacker.as_ref() {
+                contributed_stackers.push((stacker.clone(), stacked_amt));
+            }
             // peak at the next address in the set, and see if we need to sum
-            while addresses.last().map(|x| &x.0) == Some(&address) {
-                let (_, additional_amt) = addresses
+            while addresses.last().map(|x| &x.reward_address) == Some(&address) {
+                let next_contrib = addresses
                     .pop()
                     .expect("BUG: first() returned some, but pop() is none.");
+                let additional_amt = next_contrib.amount_stacked;
+
+                if let Some(stacker) = next_contrib.stacker {
+                    contributed_stackers.push((stacker.clone(), additional_amt));
+                }
+
                 stacked_amt = stacked_amt
                     .checked_add(additional_amt)
                     .expect("CORRUPTION: Stacker stacked > u128 max amount");
             }
             let slots_taken = u32::try_from(stacked_amt / threshold)
                 .expect("CORRUPTION: Stacker claimed > u32::max() reward slots");
             info!(
-                "Slots taken by {} = {}, on stacked_amt = {}, threshold = {}",
-                &address.clone(),
-                slots_taken,
-                stacked_amt,
-                threshold
+                "Reward slots taken";
+                "reward_address" => %address,
+                "slots_taken" => slots_taken,
+                "stacked_amt" => stacked_amt,
+                "pox_threshold" => threshold,
             );
             for _i in 0..slots_taken {
                 test_debug!("Add to PoX reward set: {:?}", &address);
                 reward_set.push(address.clone());
             }
+            // if stacker did not qualify for a slot *and* they have a stacker
+            //   pointer set by the PoX contract, then add them to auto-unlock list
+            if slots_taken == 0 && !contributed_stackers.is_empty() {
```

Comment:
So to be clear, this means that the following behaviors are expected?

* Alice stacks enough STX for 3 reward slots for 12 cycles.  In cycle 3, the stacking minimum rises such that she only qualifies for 1.99 reward slots (so, just 1).  Alice's remaining 0.99 slots' STX remain locked, right?

* Alice stacks enough STX for 3 reward slots for 12 cycles.  In cycle 6, she does not clear the stacking minimum.  All of her tokens unlock at cycle 6 then, right?

Reply:
Yep